### PR TITLE
Improve pagination

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ✔️

### CODE REVIEW

File: src/BaseListView.php [hunk 1]
```diff
@@ -50,8 +50,8 @@ abstract class BaseListView extends Widget
     private array $emptyTextAttributes = [];
     private string $header = '';
     private array $headerAttributes = [];
-    private string $layout = "{header}\n{toolbar}\n{items}\n{summary}\n{pager}";
-    private string $pagination = '';
+    private string $layout = "{header}\n{toolbar}\n{summary}\n{items}\n{pager}";
```
Comment (samdark) [inline src/BaseListView.php:53]: Usually summary goes under items:

```
Blog comments
[Select all] [Approve] [Reject]

Comment1
Comment2
Comment3
Comment4
Comment5

5 items out of 100
[1] 2 3 4 5 ... 20
```

File: src/BaseListView.php [hunk 2]
```diff
@@ -420,7 +428,36 @@ private function renderPagination(): string
...
+        if (!$dataReader->isPaginationRequired()) {
+            return '';
+        }
+
+        if (is_string($this->pagination)) {
+            return $this->pagination;
+        }
+
+        $pagination = $this->pagination ??
+            $dataReader instanceof KeysetPaginator ? KeysetPagination::widget() : OffsetPagination::widget();
```
Comment (samdark) [inline src/BaseListView.php:440]: Should we check for `OffsetableDataInterface` and `LimitableDataInterface`?

File: src/BasePagination.php [hunk 3]
```diff
@@ -4,17 +4,33 @@
...
+     * @psalm-var UrlCreator|null
+     */
+    private $urlCreator;
+    private string $pageParameterName = 'page';
+    private string $pageSizeParameterName = 'pagesize';
+
+    /**
+     * @psalm-var UrlParameterPlace::*
+     */
+    private int $pageParameterPlace = UrlParameterPlace::QUERY;
```
Comment (samdark) [inline src/BasePagination.php:25]:
```suggestion
    private int $pageParameterType = UrlParameterPlace::QUERY;
```

or

```suggestion
    private int $pageParameterSource = UrlParameterPlace::QUERY;
```

File: src/BasePagination.php [hunk 4]
```diff
@@ -320,28 +328,18 @@ public function urlQueryParameters(array $value): static
...
+            new PageContext(
+                $page,
+                $this->getPaginator()->getPageSize(),
+                $this->pageParameterName,
+                $this->pageSizeParameterName,
+                $this->pageParameterPlace,
+                $this->pageSizeParameterPlace,
+                $this->queryParameters,
+            )
+        );
```
Comment (samdark) [inline src/BasePagination.php:342]: This way we avoid extra call in default case:

```suggestion
        $context = new PageContext(
            $page,
            $this->getPaginator()->getPageSize(),
            $this->pageParameterName,
            $this->pageSizeParameterName,
            $this->pageParameterPlace,
            $this->pageSizeParameterPlace,
            $this->queryParameters,
        );

        if (is_callable($this->urlCreator)) {
            return $this->urlCreator($context);
        }
        
        return '#' . $context->page;
```

File: src/BaseListView.php (hunk 1 above)
Comment (vjik) [reply src/BaseListView.php]: done

File: src/BasePagination.php (hunk 3 above)
Comment (vjik) [reply src/BasePagination.php]: Renamed to "type"

File: src/BaseListView.php (hunk 2 above)
Comment (vjik) [reply src/BaseListView.php]: done

File: src/BasePagination.php (hunk 4 above)
Comment (vjik) [reply src/BasePagination.php]: done

**Pull request merged**